docs/proposals/diamond-design.md

```diff
@@ -0,0 +1,458 @@
+# Diamond workflow support
+
+Diamond workflows allow for a pipeline to scale out the generation of a single bundle.
+
+A diamond is composed of multiple splits; each split tries to write a set of files. Ideally, 2 splits do not write the
+same file.
+
+To elaborate, a bundle consists of a filesystem namespace with files in folders; each split can write files to any
+folder but should not try to write the same file.
+
+The individual nodes add files to splits without any co-ordination between them.
+
+Commit logic handles any conflicts that exist. Conflicts can only occur if the same file in the namespace of the bundle is written to by more than one split with different data.
+
+Diamonds and splits are essentially transitory objects, which live in the metadata namespace of a datamon repo.
+
+
+## Workflow
+
+The diamond workflow consists of the following steps.
+
+1. For a diamond to start, datamon needs some unique ID. Datamon SDK allows for the generation of such a unique ID for a diamond.
+
+   1. Prerequisite: generate a unique KSUID and a signalling metadata object that allows for starting splits.
+   2. All clients participating to the same diamond are expected to use the same ID at each step. **This is a requirement on client coordination**
+   3. The {diamond-id} is only used once: users cannot initialize the same diamond twice.
+   4. Initialization command, returning an identifier which is guaranteed to be unique:
```

Comment:
I thought we had decided to let the client specify an ID but we would fail initialization if the ID was already used for the repo.

Reply:
Ok. Can do